File: goimp/components/factories.go
package components

type Factory interface {
	MakeWall() Wall
	MakeDoor(Room, Room) Door
	MakeRoom(int) Room
}

type MazeFactory struct{}

func (m MazeFactory) MakeWall() Wall {
	return createSimpleWall()
}

func (m MazeFactory) MakeRoom(n int) Room {
	return createSimpleRoom(n)
}

func (m MazeFactory) MakeDoor(r1, r2 Room) Door {
	return createSimpleDoor(r1, r2)
}

type EnchantedMazeFactory struct{}

func (m EnchantedMazeFactory) MakeWall() Wall {
	return createSimpleWall()
}

func (m EnchantedMazeFactory) MakeRoom(n int) Room {
	return createEnchantedRoom(n, Spell{word: "magic~"})
}

func (m EnchantedMazeFactory) MakeDoor(r1, r2 Room) Door {
	return createSpellDoor(r1, r2)
}

type BombedMazeFactory struct{}

func (m BombedMazeFactory) MakeWall() Wall {
	return createBombedWall()
}

func (m BombedMazeFactory) MakeRoom(n int) Room {
	return createRoomWithBomb(n)
}

func (m BombedMazeFactory) MakeDoor(r1, r2 Room) Door {
	return createSimpleDoor(r1, r2)
}

File: py/patterns/components/factories.py
from .spells import Spell
from .walls import Wall, BombedWall
from .rooms import Room, EnchantedRoom, RoomWithBomb
from .doors import Door, DoorNeedingSpell


class MazeFactory:
    def make_wall(self) -> Wall:
        return Wall()

    def make_room(self, n: int) -> Room:
        return Room(n)

    def make_door(self, r1: Room, r2: Room) -> Door:
        return Door(r1, r2)


class EnchantedMazeFactory(MazeFactory):
    def make_wall(self) -> Wall:
        return Wall()

    def make_room(self, n: int) -> Room:
        return EnchantedRoom(n)

    def make_door(self, r1: Room, r2: Room) -> Door:
        return DoorNeedingSpell(r1, r2, Spell("magic word"))


class BombedMazeFactory(MazeFactory):
    def make_wall(self) -> Wall:
        return BombedWall()

    def make_room(self, n: int) -> Room:
        return RoomWithBomb(n)

    def make_door(self, r1: Room, r2: Room) -> Door:
        return Door(r1, r2)

